### PR TITLE
Sheldon build details: upgrade to shel-bot, and embedded shel-gem

### PR DESCRIPTION
**Pull request by retorquere**
The PR hosts two parts of the upgrade to Sheldon to post build details on github:

* A new endpoint `details`,  where the build details can be PUT to
* A gem, which bundles an executable script (`sheldon`), which posts the build details to that endpoint

The `styles` and `locales` repo will have to call the latter to activate this; PRs for those are incoming. Shel-bot is backwards compatible, so this should be safe to roll out.

My apologies for the long commit list but the most efficient way for me  to test was to commit and push to heroku, and I needed a lot of tests before I found a way that worked.

### CODE REVIEW

File: templates/styles/build_passed.md.erb [hunk 1]
```diff
@@ -1 +1,3 @@
 :smiley: Your submission passed all our automated tests.
+
```
Comment (adam3smith) [inline templates/styles/build_passed.md.erb]: This is used for locales and styles, right? I'd like to add something to the extent of: 

> Below are some sample citations generated by your proposed changes to citation styles

But that won't work for locales (where we don't produce output, right?). Can we differentiate? Any other ideas? This doesn't have to be perfect, just shouldn't be confusing.

**Comment by adam3smith** [comment]
@rmzelle  let me know how involved you want to be with the review of this?

File: templates/styles/build_passed.md.erb (hunk 1 above)
Comment (retorquere) [reply templates/styles/build_passed.md.erb]: No, `styles` and `locales` have their own build templates, each can be edited separately.

Comment (adam3smith) [reply templates/styles/build_passed.md.erb]: I could have figured that out from the path...

**Comment by rmzelle** [comment]
Looks good to me (for as far as I can judge this). We can always tweak the template files after the initial merge.

@retorquere, do you know if it's enough to merge the current PR, or do we need to deploy the new version of Sheldon to Heroku as well? I don't think I was ever involved in the latter.

**Comment by adam3smith** [comment]
Dang, messed up the squash on merge :( . Probably not much of an issue here. I'll be more careful on styles

**Comment by retorquere** [comment]
Didn't squash & merge just work?

Sheldon must be redeployed to heroku, but that may have already happened as there's a Travis deploy stanza. Has the memcached addon be added? Sheldon won't work without it active. If it wasn't, it can still be added, but Sheldon likely needs to be restarted to pick it up.

**Comment by retorquere** [comment]
I can confirm that sheldon has been deployed, the current version that's running is 0.1.33. I can't readily tell whether the memcached addon has been added though.

**Comment by adam3smith** [comment]
Any idea how I could check for memcached?

**Comment by retorquere** [comment]
Someone who has access to the heroku project that runs it can see if it's enabled.

**Comment by adam3smith** [comment]
@rmzelle Do you? -- I don't even have a heroku account. Or @inukshuk could you check?

**Comment by rmzelle** [comment]
> I can't readily tell whether the memcached addon has been added though.

I'm a collaborator on the Heroku shelbot app (with Sylvester as the owner). When I look at https://dashboard.heroku.com/apps/shelbot/resources, no add-ons are installed/active. Searching for "mem" in the add-on directory shows there is https://elements.heroku.com/addons/memcachier and https://elements.heroku.com/addons/memcachedcloud. Do you need one of these?

**Comment by retorquere** [comment]
The 2nd one

**Comment by rmzelle** [comment]
Just installed (at the free tier):

image

**Comment by adam3smith** [comment]
(we discussed that the free tier should be more than enough for our needs).

@rmzelle Could you invite me to the team? using my karcher@u.... email address for Heroku

**Comment by retorquere** [comment]
Alright -- if Sheldon has been restarted since, it will pick up that instance from now on, and the `styles` and `locales` PRs would then be ready for review and merge.

**Comment by rmzelle** [comment]
> if Sheldon has been restarted since

Not sure that happened yet. There is an option to "Restart all dynos" that might do the trick, or I guess we could just make a trivial commit to this repo to trigger a new deployment.

(we're also on the old cedar-14 stack, by the way, which has just been end-of-lifed: https://devcenter.heroku.com/changelog-items/1603. We should probably upgrade to heroku-18, but I don't know if that will break anything)

**Comment by retorquere** [comment]
If you have the heroku cli installed, `heroku restart` in the Sheldon directory, or `heroku restart -a shelbot` elsewhere, should do the trick.

I've been doing my tests on `heroku-18` and I didn't experience any problems.

**Comment by retorquere** [comment]
Oh sorry, missed that part: yes, `restart all dynos` will do it. Sheldon has only a single dyno probably, so the language there might be confusing, but that's basically it.

you can also use the `heroku logs` command to see if sheldon was wound down already. If it's wound down, new requests should spin it up afresh. But an explicit restart seems best.

**Comment by rmzelle** [comment]
Done.

**Pull request merged**